File: src/gesture/DoublePinchDetector.ts
import { GestureType } from '../types';

/**
 * Universal Double Pinch Detector for Play/Pause.
 *
 * Grounded in official Wear OS motion sensor mechanics:
 * A pinch gesture (index finger snapping against the thumb) transmits a sharp,
 * transient micro-acceleration shockwave into the watch chassis.
 *
 * Universal & Tilt-Invariant Design:
 * - Employs 3D acceleration magnitude (along with dedicated Z-axis sensitivity)
 *   so the gesture works when the watch is face-up, tilted, or hanging down.
 * - Enforces quiet gyroscope state (to avoid confusion with wrist rotations or arm swings).
 * - Tracks a robust two-pinch state machine with rebound/re-arm cycles.
 */
export class DoublePinchDetector {
  private thresholdUp: number;
  private thresholdDown: number;
  private maxAccMag: number;
  private windowMs: number;
  private cooldownMs: number;
  private maxGyroMagnitude: number;

  private static readonly PINCH_TIMEOUT_MS = 450;
  private static readonly MIN_PINCH_INTERVAL_MS = 100;

  private lastGestureTime = 0;
  private pinches: number[] = [];

  private state: 'IDLE' | 'IMPULSE_ACTIVE' = 'IDLE';
  private impulseStartTime = 0;
  private armed = true;
  private stableStartTime = 0;

  constructor(
    thresholdUp = 3.2,
    thresholdDown = -1.8,
    maxAccMag = 14.0,
    windowMs = 900,
    cooldownMs = 1000,
    maxGyroMagnitude = 2.5
  ) {
    this.thresholdUp = thresholdUp;
    this.thresholdDown = thresholdDown;
    this.maxAccMag = maxAccMag;
    this.windowMs = windowMs;
    this.cooldownMs = cooldownMs;
    this.maxGyroMagnitude = maxGyroMagnitude;
  }

  public updateSettings(pinchThreshold: number, cooldownMs: number) {
    this.thresholdUp = pinchThreshold;
    this.thresholdDown = -(pinchThreshold * 0.6);
    this.cooldownMs = cooldownMs;
  }

  public process(
    timestamp: number,
    gyroX: number,
    gyroY: number,
    gyroZ: number,
    linAccX: number,
    linAccY: number,
    linAccZ: number
  ): GestureType | null {
    const accMag = Math.sqrt(linAccX * linAccX + linAccY * linAccY + linAccZ * linAccZ);

    // 1. Rearm check after firing
    if (!this.armed) {
      if (accMag <= 1.8) {
        if (this.stableStartTime === 0) this.stableStartTime = timestamp;
        if (
          timestamp - this.stableStartTime >= 80 &&
          timestamp - this.lastGestureTime >= this.cooldownMs
        ) {
          this.armed = true;
          this.stableStartTime = 0;
        }
      } else {
        this.stableStartTime = 0;
        if (timestamp - this.lastGestureTime >= this.cooldownMs + 400) {
          this.armed = true;
        }
      }
      return null;
    }

    // 2. Cooldown check
    if (timestamp - this.lastGestureTime < this.cooldownMs) return null;

    // 3. Gyroscope guard: pinch must occur with low rotational movement
    const gyroMag = Math.max(Math.abs(gyroX), Math.abs(gyroY), Math.abs(gyroZ));
    if (gyroMag > this.maxGyroMagnitude) {
      if (
        this.state === 'IMPULSE_ACTIVE' &&
        timestamp - this.impulseStartTime > DoublePinchDetector.PINCH_TIMEOUT_MS
      ) {
        this.state = 'IDLE';
      }
      return null;
    }

    // 4. Clean expired pinch timestamps from window
    if (this.pinches.length > 0) {
      this.pinches = this.pinches.filter((t) => timestamp - t <= this.windowMs);
    }

    // 5. Detect sharp impulse (Z-axis or total 3D magnitude)
    const isImpulse =
      (linAccZ > this.thresholdUp && linAccZ <= this.maxAccMag) ||
      (accMag > this.thresholdUp && accMag <= this.maxAccMag);

    if (this.state === 'IDLE') {
      if (isImpulse) {
        this.state = 'IMPULSE_ACTIVE';
        this.impulseStartTime = timestamp;
      }
    } else if (this.state === 'IMPULSE_ACTIVE') {
      // Rebound or return to baseline within timeout window
      const isRebound =
        linAccZ < this.thresholdDown ||
        (timestamp - this.impulseStartTime >= 40 && accMag < this.thresholdUp * 0.5);

      if (isRebound) {
        this.pinches.push(timestamp);
        this.state = 'IDLE';

        // Check if we have two valid pinches
        if (this.pinches.length >= 2) {
          const first = this.pinches[this.pinches.length - 2];
          const second = this.pinches[this.pinches.length - 1];
          const interval = second - first;

          if (interval >= DoublePinchDetector.MIN_PINCH_INTERVAL_MS && interval <= this.windowMs) {
            this.lastGestureTime = timestamp;
            this.pinches = [];
            this.armed = false;
            this.stableStartTime = 0;
            return GestureType.PLAY_PAUSE;
          }
        }
      } else if (timestamp - this.impulseStartTime > DoublePinchDetector.PINCH_TIMEOUT_MS) {
        this.state = 'IDLE';
      }
    }

    return null;
  }

  public reset() {
    this.pinches = [];
    this.state = 'IDLE';
    this.impulseStartTime = 0;
    this.lastGestureTime = 0;
    this.armed = true;
    this.stableStartTime = 0;
  }
}

File: app/src/main/java/com/example/gesturemusicwear/DoublePinchDetector.java
package com.example.gesturemusicwear;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Double pinch detector for play/pause.
 *
 * Uses totalG (raw accelerometer magnitude in g-force) with a two-pinch FSM.
 * Requires two valid pinch impulses within a time window to fire,
 * preventing single-impulse false triggers from taps or bumps.
 * Gyroscope guard rejects during wrist rotation.
 *
 * Improved sensitivity with relaxed thresholds for better detection.
 */
public class DoublePinchDetector {
    public static final int RESULT_NONE = 0;
    public static final int RESULT_PLAY_PAUSE = 1;

    private static final long PINCH_TIMEOUT_MS = 500; // increased from 450ms
    private static final long MIN_PINCH_INTERVAL_MS = 80;  // reduced from 100ms

    private float thresholdUp;   // g-force
    private float thresholdDown; // g-force (negative)
    private float maxTotalG;
    private long windowMs;
    private long cooldownMs;
    private float maxGyroMagnitude; // rad/s

    private long lastGestureTime = 0;
    private final ArrayList<Long> pinches = new ArrayList<>();

    private static final int STATE_IDLE = 0;
    private static final int STATE_IMPULSE_ACTIVE = 1;
    private int state = STATE_IDLE;
    private long impulseStartTime = 0;
    private boolean armed = true;
    private long stableStartTime = 0;

    public DoublePinchDetector(float thresholdUp, float thresholdDown, float maxTotalG,
                               long windowMs, long cooldownMs, float maxGyroMagnitude) {
        this.thresholdUp = thresholdUp;
        this.thresholdDown = thresholdDown;
        this.maxTotalG = maxTotalG;
        this.windowMs = windowMs;
        this.cooldownMs = cooldownMs;
        this.maxGyroMagnitude = maxGyroMagnitude;
    }

    public void updateSettings(float pinchThreshold, long cooldownMs) {
        this.thresholdUp = pinchThreshold;
        this.thresholdDown = -(pinchThreshold * 0.6f);
        this.cooldownMs = cooldownMs;
    }

    public int process(long timestamp, float gx, float gy, float gz,
                       float accX, float accY, float accZ) {
        float accMag = (float) Math.sqrt(accX * accX + accY * accY + accZ * accZ);
        float totalG = accMag / 9.80665f;

        if (!armed) {
            if (Math.abs(totalG - 1.0f) <= 0.3f) {
                if (stableStartTime == 0) stableStartTime = timestamp;
                if (timestamp - stableStartTime >= 80 && timestamp - lastGestureTime >= cooldownMs) {
                    armed = true;
                    stableStartTime = 0;
                }
            } else {
                stableStartTime = 0;
                if (timestamp - lastGestureTime >= cooldownMs + 400) {
                    armed = true;
                }
            }
            return RESULT_NONE;
        }

        if (timestamp - lastGestureTime < cooldownMs) return RESULT_NONE;

        float gyroMag = Math.max(Math.abs(gx), Math.max(Math.abs(gy), Math.abs(gz)));
        if (gyroMag > maxGyroMagnitude) {
            if (state == STATE_IMPULSE_ACTIVE && timestamp - impulseStartTime > PINCH_TIMEOUT_MS) {
                state = STATE_IDLE;
            }
            return RESULT_NONE;
        }

        Iterator<Long> it = pinches.iterator();
        while (it.hasNext()) {
            if (timestamp - it.next() > windowMs) it.remove();
        }

        boolean isImpulse = totalG >= thresholdUp * 0.9f && totalG <= maxTotalG; // reduced threshold by 10% for better sensitivity

        if (state == STATE_IDLE) {
            if (isImpulse) {
                state = STATE_IMPULSE_ACTIVE;
                impulseStartTime = timestamp;
            }
        } else if (state == STATE_IMPULSE_ACTIVE) {
            boolean isRebound = totalG < Math.abs(thresholdDown)
                || (timestamp - impulseStartTime >= 40 && totalG < thresholdUp * 0.5f);

            if (isRebound) {
                pinches.add(timestamp);
                state = STATE_IDLE;

                if (pinches.size() >= 2) {
                    long first = pinches.get(pinches.size() - 2);
                    long second = pinches.get(pinches.size() - 1);
                    long interval = second - first;

                    if (interval >= MIN_PINCH_INTERVAL_MS && interval <= windowMs) {
                        lastGestureTime = timestamp;
                        pinches.clear();
                        armed = false;
                        stableStartTime = 0;
                        return RESULT_PLAY_PAUSE;
                    }
                }
            } else if (timestamp - impulseStartTime > PINCH_TIMEOUT_MS) {
                state = STATE_IDLE;
            }
        }

        return RESULT_NONE;
    }

    public void reset() {
        pinches.clear();
        state = STATE_IDLE;
        impulseStartTime = 0;
        lastGestureTime = 0;
        armed = true;
        stableStartTime = 0;
    }
}